libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m

```diff
@@ -292,29 +321,40 @@ - (void)enqueueRequest:(SFRestRequest *)request requestDelegate:(id<SFRestReques
             if (error) {
                 [SFSDKCoreLogger d:[strongSelf class] format:@"REST request failed with error: Error Code: %ld, Description: %@, URL: %@", (long) error.code, error.localizedDescription, finalRequest.URL];
                 id dataForDelegate = [strongSelf prepareDataForDelegate:data request:request response:response];
-                [strongSelf notifyDelegateOfFailure:requestDelegate request:request data:dataForDelegate rawResponse:response error:error];
+                if (failureBlock) {
+                    failureBlock(dataForDelegate, error, response);
+                }
+                [self removeActiveRequestObject:request];
                 return;
             }
-
+            
             // Timeout.
             if (!response) {
-                [strongSelf notifyDelegateOfFailure:requestDelegate request:request data:nil rawResponse:nil error:nil];
+                if (failureBlock) {
+                    failureBlock(nil, nil, nil);
+                }
+                [self removeActiveRequestObject:request];
                 return;
             }
             NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];
 
             // 2xx indicates success.
             if ([SFRestAPI isStatusCodeSuccess:statusCode]) {
                 id dataForDelegate = [strongSelf prepareDataForDelegate:data request:request response:response];
-                [strongSelf notifyDelegateOfSuccess:requestDelegate request:request data:dataForDelegate rawResponse:response];
+                if (successBlock) {
+                    successBlock(dataForDelegate, response);
```

Comment:
Is `[self removeActiveRequestObject:request];` needed for the success case too?

Reply:
you might be right.  Let me see if I can move [self removeActiveRequestObject:request]; up a level in the callback so we know it's not getting missed anywhere.

Reply:
just saw the new way and it looks way cleaner!